### PR TITLE
avoid copying data when not reading genotypes

### PR DESCRIPTION
**Pull request by danking**
@tpoterba you were in here recently for performance so your eyes are appreciated.

I simplified things a bit and localized almost all the parsing logic to `BgenRecord`. The contract for `advance` is that it is always called when `bfis` is pointing at the start of a record _or_ at or past the `end`. Advance will return the position to the start of a record or at or past the `end`. It returns true if there was a new record found. False otherwise.

I avoided a couple allocating patterns.

The rest of the diffs are copy pastes and some indentation changes.

### CODE REVIEW

**Comment by danking** [comment]
I made a tiny change to the alleles check too. Look at https://github.com/hail-is/hail/pull/3783/files?w=1

**Comment by danking** [comment]
Also @cseed @jigold your thoughts on this small refactor appreciated if you have time.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 1]
```diff
@@ -1,149 +1,203 @@
...
-        fatal(s"Ploidy value must equal to 2. Found $ploidy.")
-      i += 1
+      val nExpectedBytesProbs = (nSamples * (nGenotypes - 1) * nBitsPerProb + 7) / 8
+      if (reader.length != nExpectedBytesProbs + nSamples + 10)
+        fatal(s"""Number of uncompressed bytes `${ reader.length }' does not
+                 |match the expected size `$nExpectedBytesProbs'.""".stripMargin)
+    } else {
+      bfis.skipBytes(dataSize)
     }
-    if (i != nSamples)
```
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: Why did you delete this check?

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 2]
```diff
@@ -1,149 +1,203 @@
...
+      if (phase != 0 && phase != 1)
+        fatal(s"Value for phase must be 0 or 1. Found $phase.")
+      val isPhased = phase == 1
 
-    val nAlleles = reader.readShort()
-    if (nAlleles != expectedNumAlleles)
-      fatal(s"""Value for `nAlleles' in genotype probability data storage is
-               |not equal to value in variant identifying data. Expected
-               |$expectedNumAlleles but found $nAlleles.""".stripMargin)
-    if (nAlleles != 2)
```
Comment (danking) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: pulled this check up to the earlier read of `nAlleles` (which is stored twice for some reason??)

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala (hunk 1 above)
Comment (jigold) [reply src/main/scala/is/hail/io/bgen/BgenRecord.scala]: No, I meant the sample check here that the number of ploidy values read is the expected number of samples. And yes, the nAlleles is stored in the file twice.
